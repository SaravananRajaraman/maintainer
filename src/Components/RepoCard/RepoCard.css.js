import CardHeader from '@material-ui/core/CardHeader';
import styled from 'styled-components';

export const StyledHeader = styled(CardHeader)`
    &&{
        padding-top : 2px;
        padding-bottom : 2px;
        background-image: linear-gradient(to right, #314755 0%, #26a0da 100%);
        // background-image: linear-gradient(to right, #003973 0%, #E5E5BE 100%);
        color : #fff;
        font-size : 34px;        
    }`   

export const StyledBox = styled('div')`
        align-items : center;
        display : flex;      
        margin-top : ${props => props.margin ? '10px' : '0'};
`
export const IconBox = styled('div')`
        display: inline-grid;
        justify-items: center; 
        width: 100px;    
`