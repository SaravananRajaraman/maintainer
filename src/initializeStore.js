import { createStore,applyMiddleware } from "redux";
import thunk from 'redux-thunk';
import createReducer from "./reducers";

let middleware = [ thunk ];

const initializeStore = () => {
  
  const store = createStore(    
    createReducer(),       
    applyMiddleware(...middleware)     
  );   
  return store;
};

export default initializeStore;