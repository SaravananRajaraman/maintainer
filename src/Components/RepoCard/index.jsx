import React from 'react';
import {Card,CardContent,Typography,Divider,Tooltip,Chip,Snackbar,Badge,Link} from '@material-ui/core';
import FolderOpenIcon from '@material-ui/icons/FolderOpen';
import PersonIcon from '@material-ui/icons/Person';
import LinkIcon from '@material-ui/icons/Link';
import {StyledHeader,StyledBox,IconBox} from  './RepoCard.css';
import FolderSpecialIcon from '@material-ui/icons/FolderSpecial';
import AccountTreeIcon from '@material-ui/icons/AccountTree';
import MuiAlert from '@material-ui/lab/Alert';
import AppsIcon from '@material-ui/icons/Apps';
import PagesIcon from '@material-ui/icons/Pages';
import ExtensionIcon from '@material-ui/icons/Extension';
import Skeleton from '@material-ui/lab/Skeleton';

const Types = ['Component','Module','Application'];

export const RepoSkeleton = () => {return(

            <React.Fragment>
                <Card elevation={6}>
                <StyledHeader 
                            avatar={<Skeleton animation="wave" variant="circle" width={40} height={40} />}
                            title={<Skeleton animation="wave" height={30} width="80%" />}
                            />                           
                <CardContent>
                    <StyledBox>
                        <IconBox>
                            <LinkIcon/>                    
                            <Typography variant={'caption'}  >
                                Link
                            </Typography>
                        </IconBox>
                        <Skeleton animation="wave" height={30} width={180} />                                              
                    </StyledBox>     
                    <Divider  />
                    <StyledBox margin>
                        <IconBox>
                            <PersonIcon/>                    
                            <Typography variant={'caption'}  >
                                Owner
                            </Typography>
                        </IconBox>                        
                        <div >
                            {[0,1].map(name=>                                                                  
                                        <Chip key={name} style={{padding:'4px', margin : '4px'}} 
                                        icon={<Skeleton animation="wave" variant="circle" width={20} height={20}/>}                                         
                                        label={<Skeleton animation="wave" height={20} width={80}/>} variant="outlined"/>                                                                   
                            )}                            
                        </div>
                    </StyledBox>    
                    <Divider />
                    <StyledBox margin>
                        <IconBox>
                            <PersonIcon/>                    
                            <Typography variant={'caption'}  >
                                Reviewer
                            </Typography>
                        </IconBox>                                                
                        <div >
                            {[0,1].map(name=>                                                                  
                                        <Chip key={name} style={{padding:'4px', margin : '4px'}} 
                                        icon={<Skeleton animation="wave" variant="circle" width={20} height={20}/>}                                         
                                        label={<Skeleton animation="wave" height={20} width={80}/>} variant="outlined"/>                                                                
                            )} 
                        </div>
                    </StyledBox>   
                    <Divider />
                    <StyledBox margin>
                        <IconBox>
                            <FolderSpecialIcon/>                    
                            <Typography variant={'caption'}  >
                                Repo Type
                            </Typography>
                        </IconBox>                        
                        <div >                        
                            <StyledBox>                                           
                                    <Skeleton animation="wave" variant="circle" width={30} height={30}/>                
                                    <Skeleton animation="wave" height={30} width={100} style={{marginLeft : '4px'}}/>
                            </StyledBox>                                                                                       
                        </div>
                    </StyledBox>  
                    <Divider />
                    <StyledBox margin>
                        <IconBox>
                            <AccountTreeIcon/>                    
                            <Typography variant={'caption'}  >                                
                                Dependency
                            </Typography>
                        </IconBox>                        
                        <div >
                            {[0,1].map(dependency =>                                
                                    <Badge key={dependency} color="secondary" badgeContent={''} style={{margin : '15px'}}>
                                        <Skeleton animation="wave" height={30} width={100}/>
                                    </Badge>                                                                                         
                            )}                        
                        </div>
                    </StyledBox>                                                                                  
                </CardContent>                
            </Card>            
        </React.Fragment>
)}           

function getIcon(name){
        let index = Types.indexOf(name);
        if(index === 0){
            return <ExtensionIcon color={'action'}/>
        }
        else if(index === 1){
            return <PagesIcon color={'action'}/>
        }
        else{
            return <AppsIcon color={'action'}/>
        }

}

function Alerts(props) {
                return <MuiAlert elevation={6} variant="filled" {...props} />;
              }            

const copyToClipboard = str => {
                const el = document.createElement('textarea');
                el.value = str;
                el.setAttribute('readonly', '');
                el.style.position = 'absolute';
                el.style.left = '-9999px';
                document.body.appendChild(el);
                el.select();
                document.execCommand('copy');
                document.body.removeChild(el);
              };            

export default function RepoCard({Repo}){

    const [open, setOpen] = React.useState(false);

    const handleClick = () => setOpen(!open);    
  
    const copyText = (str) => {copyToClipboard(str); handleClick();};

    return(
        <React.Fragment >
             <Card elevation={6} data-test='RepoCard-Component'>
                <StyledHeader title={Repo.name} 
                            avatar={<FolderOpenIcon size='medium'/>}
                            titleTypographyProps = {{variant : 'h6'}}                
                />                        
                <CardContent>
                    <StyledBox>
                        <IconBox>
                            <LinkIcon/>                    
                            <Typography variant={'caption'}  >
                                Link
                            </Typography>
                        </IconBox>
                        <Link href={Repo.link} target="_blank" rel="noopener">
                            {Repo.link}
                        </Link>                                                
                    </StyledBox>     
                    <Divider  />
                    <StyledBox margin>
                        <IconBox>
                            <PersonIcon/>                    
                            <Typography variant={'caption'}  >
                                Owner
                            </Typography>
                        </IconBox>                        
                        <div >
                            {Repo.owner.map(name=>                              
                                    <Tooltip key={name.id} title={name.emailId}>
                                        <Chip style={{padding:'4px', margin : '4px'}} 
                                        icon={<PersonIcon/> } 
                                        onClick={()=>copyText(name.emailId)} 
                                        label={name.name} variant="outlined"/>                            
                                    </Tooltip>    
                            )}                            
                        </div>
                    </StyledBox>    
                    <Divider />
                    <StyledBox margin>
                        <IconBox>
                            <PersonIcon/>                    
                            <Typography variant={'caption'}  >
                                Reviewer
                            </Typography>
                        </IconBox>                                                
                        <div >
                            {Repo.reviewer.map(name=>                              
                                    <Tooltip key={name.id} title={name.emailId}>
                                        <Chip style={{padding:'4px', margin : '4px'}} 
                                        icon={<PersonIcon/> } 
                                        onClick={()=>copyText(name.emailId)} 
                                        label={name.name} variant="outlined"/>                                                                    
                                    </Tooltip>    
                            )}
                        </div>
                    </StyledBox>   
                    <Divider />
                    <StyledBox margin>
                        <IconBox>
                            <FolderSpecialIcon/>                    
                            <Typography variant={'caption'}  >
                                Repo Type
                            </Typography>
                        </IconBox>                        
                        <div >
                        {Repo.repoType && 
                            <StyledBox>                                           
                                    {getIcon(Repo.repoType)}                
                                    <Typography style={{marginLeft : '4px'}} variant={'h6'} color={'textSecondary'}>{Repo.repoType}</Typography>                                                                    
                            </StyledBox>
                        }                                                                 
                        </div>
                    </StyledBox>  
                    <Divider />
                    <StyledBox margin>
                        <IconBox>
                            <AccountTreeIcon/>                    
                            <Typography variant={'caption'}  >                                
                                Dependency
                            </Typography>
                        </IconBox>                        
                        <div >
                            {Repo.dependency.map(dependency =>                                
                                    <Badge key={dependency.package} color="secondary" badgeContent={dependency.version} style={{margin : '15px'}}>
                                        <Typography variant={'h6'} color={'textSecondary'}>{dependency.package}</Typography>
                                    </Badge>                                                                                         
                            )}                        
                        </div>
                    </StyledBox>  
                    <Snackbar open={open} autoHideDuration={1000} onClose={handleClick}>
                            <Alerts onClose={handleClick} severity="success">
                                Email copied to clipboard
                            </Alerts>
                    </Snackbar>                                                             
                </CardContent>                
            </Card>            
        </React.Fragment>
    )
}