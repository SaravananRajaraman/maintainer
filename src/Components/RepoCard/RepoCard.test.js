import React from 'react';
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import RepoCard from './index';
import {findTestAttr} from '../../Utils';
import initializeStore from '../../initializeStore';

const store = initializeStore();

const mock = {
    "name"  : "Repo 2",
    "link"  : "http://www.Repo2.com",
    "owner" : [
                {
                    "id" : 5,
                    "emailId" : "owner5@gmail.com",
                    "name" : "owner5"
                },
                {
                    "id" : 2,
                    "emailId" : "owner2@gmail.com",
                    "name" : "owner2"
                }
            ],
    "reviewer" : [
                {
                    "id" : 8,
                    "emailId" : "reviewer8@gmail.com",
                    "name" : "reviewer8"
                },
                {
                    "id" : 3,
                    "emailId" : "reviewer3@gmail.com",
                    "name" : "reviewer3"
                }
            ],
    "repoType" : "Application",
    "dependency" : [
                {
                    "package" : "package3",
                    "version" : "3.0.0"
                },
                {
                    "package" : "package2",
                    "version" : "2.0.0"
                }
            ]
};

const setUp = (props={}) => {         
    const component = shallow(<RepoCard store={store} Repo={mock}/>);        
    return component;
}

describe('RepoCard Component', ()=>{  

    describe('Should test the snapshot and render functionality',()=>{

        let component;

        beforeEach(()=>{
            component = setUp();
        });

        it('Should render the component without error',()=>{                      
            let wrapper = findTestAttr(component,'RepoCard-Component');
            expect(wrapper.length).toBe(1); 
        });

        it('Should match the RepoCard',()=>{                              
            expect(toJson(component)).toMatchSnapshot(); 
        });       

    });

    describe('Should test the event functions',()=>{
                
        it('Should click the snackbar function',()=>{    
            let setState = jest.fn();                   
            const useStateSpy = jest.spyOn(React, 'useState');
            useStateSpy.mockImplementation(() => [true, setState]);
            let component =  setUp();                                                
            let wrapper = findTestAttr(component,'SnackBar');           
            wrapper.props().onClose(); 
            expect(setState).toHaveBeenCalledWith(false);   
            jest.clearAllMocks();     
        });

        it('Should click the owner chip click function',()=>{  
            let setState = jest.fn();   
            document.execCommand = jest.fn();            
            const useStateSpy = jest.spyOn(React, 'useState');
            useStateSpy.mockImplementation(() => [false, setState]);
            let component =  setUp();                                          
            let wrapper = findTestAttr(component,'owner-5');   
            wrapper.props().onClick();             
            expect(setState).toHaveBeenCalledWith(true); 
            jest.clearAllMocks();    
        });

        it('Should click the reviewer chip click function',()=>{  
            let setState = jest.fn();   
            document.execCommand = jest.fn();            
            const useStateSpy = jest.spyOn(React, 'useState');
            useStateSpy.mockImplementation(() => [false, setState]);
            let component =  setUp();                                          
            let wrapper = findTestAttr(component,'reviewer-8');   
            wrapper.props().onClick();             
            expect(setState).toHaveBeenCalledWith(true); 
            jest.clearAllMocks();    
        });
    });

});