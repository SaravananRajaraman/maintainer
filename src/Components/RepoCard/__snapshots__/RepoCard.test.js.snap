// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoCard Component Should test the snapshot and render functionality Should match the RepoCard 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Card))
    data-test="RepoCard-Component"
    elevation={6}
  >
    <Styled(WithStyles(ForwardRef(CardHeader)))
      avatar={
        <UNDEFINED
          size="medium"
        />
      }
      title="Repo 2"
      titleTypographyProps={
        Object {
          "variant": "h6",
        }
      }
    />
    <WithStyles(ForwardRef(CardContent))>
      <styled.div>
        <styled.div>
          <LinkIcon />
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Link
          </WithStyles(ForwardRef(Typography))>
        </styled.div>
        <WithStyles(ForwardRef(Link))
          href="http://www.Repo2.com"
          rel="noopener"
          target="_blank"
        >
          http://www.Repo2.com
        </WithStyles(ForwardRef(Link))>
      </styled.div>
      <WithStyles(ForwardRef(Divider)) />
      <styled.div
        margin={true}
      >
        <styled.div>
          <PersonIcon />
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Owner
          </WithStyles(ForwardRef(Typography))>
        </styled.div>
        <div>
          <WithStyles(ForwardRef(Tooltip))
            key="5"
            title="owner5@gmail.com"
          >
            <WithStyles(ForwardRef(Chip))
              data-test="owner-5"
              icon={<UNDEFINED />}
              label="owner5"
              onClick={[Function]}
              style={
                Object {
                  "margin": "4px",
                  "padding": "4px",
                }
              }
              variant="outlined"
            />
          </WithStyles(ForwardRef(Tooltip))>
          <WithStyles(ForwardRef(Tooltip))
            key="2"
            title="owner2@gmail.com"
          >
            <WithStyles(ForwardRef(Chip))
              data-test="owner-2"
              icon={<UNDEFINED />}
              label="owner2"
              onClick={[Function]}
              style={
                Object {
                  "margin": "4px",
                  "padding": "4px",
                }
              }
              variant="outlined"
            />
          </WithStyles(ForwardRef(Tooltip))>
        </div>
      </styled.div>
      <WithStyles(ForwardRef(Divider)) />
      <styled.div
        margin={true}
      >
        <styled.div>
          <PersonIcon />
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Reviewer
          </WithStyles(ForwardRef(Typography))>
        </styled.div>
        <div>
          <WithStyles(ForwardRef(Tooltip))
            key="8"
            title="reviewer8@gmail.com"
          >
            <WithStyles(ForwardRef(Chip))
              data-test="reviewer-8"
              icon={<UNDEFINED />}
              label="reviewer8"
              onClick={[Function]}
              style={
                Object {
                  "margin": "4px",
                  "padding": "4px",
                }
              }
              variant="outlined"
            />
          </WithStyles(ForwardRef(Tooltip))>
          <WithStyles(ForwardRef(Tooltip))
            key="3"
            title="reviewer3@gmail.com"
          >
            <WithStyles(ForwardRef(Chip))
              data-test="reviewer-3"
              icon={<UNDEFINED />}
              label="reviewer3"
              onClick={[Function]}
              style={
                Object {
                  "margin": "4px",
                  "padding": "4px",
                }
              }
              variant="outlined"
            />
          </WithStyles(ForwardRef(Tooltip))>
        </div>
      </styled.div>
      <WithStyles(ForwardRef(Divider)) />
      <styled.div
        margin={true}
      >
        <styled.div>
          <FolderSpecialIcon />
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Repo Type
          </WithStyles(ForwardRef(Typography))>
        </styled.div>
        <div>
          <styled.div>
            <AppsIcon
              color="action"
            />
            <WithStyles(ForwardRef(Typography))
              color="textSecondary"
              style={
                Object {
                  "marginLeft": "4px",
                }
              }
              variant="h6"
            >
              Application
            </WithStyles(ForwardRef(Typography))>
          </styled.div>
        </div>
      </styled.div>
      <WithStyles(ForwardRef(Divider)) />
      <styled.div
        margin={true}
      >
        <styled.div>
          <AccountTreeIcon />
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Dependency
          </WithStyles(ForwardRef(Typography))>
        </styled.div>
        <div>
          <WithStyles(ForwardRef(Badge))
            badgeContent="3.0.0"
            color="secondary"
            key="package3"
            style={
              Object {
                "margin": "15px",
              }
            }
          >
            <WithStyles(ForwardRef(Typography))
              color="textSecondary"
              variant="h6"
            >
              package3
            </WithStyles(ForwardRef(Typography))>
          </WithStyles(ForwardRef(Badge))>
          <WithStyles(ForwardRef(Badge))
            badgeContent="2.0.0"
            color="secondary"
            key="package2"
            style={
              Object {
                "margin": "15px",
              }
            }
          >
            <WithStyles(ForwardRef(Typography))
              color="textSecondary"
              variant="h6"
            >
              package2
            </WithStyles(ForwardRef(Typography))>
          </WithStyles(ForwardRef(Badge))>
        </div>
      </styled.div>
      <WithStyles(ForwardRef(Snackbar))
        autoHideDuration={1000}
        data-test="SnackBar"
        onClose={[Function]}
        open={false}
      >
        <WithStyles(ForwardRef(Alert))
          elevation={6}
          onClose={[Function]}
          severity="success"
          variant="filled"
        >
          Email copied to clipboard
        </WithStyles(ForwardRef(Alert))>
      </WithStyles(ForwardRef(Snackbar))>
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(Card))>
</Fragment>
`;
